raw-ipa-lib/src/user.rs

```diff
@@ -0,0 +1,146 @@
+            return Err(Error::InvalidId);
+        }
+        Ok(v)
+    }
+
+    /// # Errors
+    /// When the file cannot be written.
+    #[cfg(feature = "enable-serde")]
+    pub fn save(&self, dir: &Path) -> Res<()> {
+        let f = self.filename(dir);
+        fs::write(f, serde_json::to_string(self)?.as_bytes())?;
+        Ok(())
+    }
+
+    fn point_from_matchkey(mk: &[u8; 32]) -> RistrettoPoint {
+        // Note that ristretto wants 64 bytes of input; also we don't know if the input is uniform.
+        // TODO: Consider salting this input somehow (with the origin, perhaps).
+        //       The caveat being that anything we do needs to be standardized.
+        RistrettoPoint::hash_from_bytes::<Sha512>(&mk[..])
+    }
+
+    pub fn set_matchkey(&mut self, origin: &str, mk: &[u8; 32]) {
+        let m = Self::point_from_matchkey(mk);
+        let emk = self.threshold_key.encrypt(m, &mut thread_rng());
+        self.encrypted_match_keys.insert(String::from(origin), emk);
+    }
+
+    /// Create an encrypted matchkey for the identified origin.
+    #[must_use]
+    pub fn encrypt_matchkey(&self, origin: &str) -> Ciphertext {
+        let mut rng = thread_rng();
+        // TODO: determine if we need to hide the timing sidechannel here.
+        if let Some(emk) = self.encrypted_match_keys.get(origin) {
+            self.threshold_key.rerandomise(*emk, &mut rng)
+        } else {
+            Ciphertext::from((
+                RistrettoPoint::random(&mut rng),
+                RistrettoPoint::random(&mut rng),
+            ))
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::User;
+    use crate::threshold::{
+        DecryptionKey as ThresholdDecryptionKey, EncryptionKey as ThresholdEncryptionKey,
+    };
+    use rand::thread_rng;
+
+    const MATCHKEY: &[u8; 32] = &[0; 32];
+    const ORIGIN: &str = "example.com";
+
+    #[test]
+    fn matchkey_two() {
+        let mut rng = thread_rng();
+        let d1 = ThresholdDecryptionKey::new(&mut rng);
+        let d2 = ThresholdDecryptionKey::new(&mut rng);
+        let tek = ThresholdEncryptionKey::new(&[d1.encryption_key(), d2.encryption_key()]);
+        let mut u = User::new(0, tek);
+        u.set_matchkey(ORIGIN, MATCHKEY);
+
+        let c = u.encrypt_matchkey(ORIGIN);
+        let partial1 = d1.threshold_decrypt(c);
+        let complete1 = d2.decrypt(partial1);
+        assert_eq!(complete1, User::point_from_matchkey(MATCHKEY));
+
+        // A redundant check that ordering doesn't matter.
+        let partial2 = d2.threshold_decrypt(c);
+        let complete2 = d1.decrypt(partial2);
+        assert_eq!(complete1, complete2);
+    }
+
+    #[test]
+    fn matchkey_random() {
+        let mut rng = thread_rng();
+        let d1 = ThresholdDecryptionKey::new(&mut rng);
+        let d2 = ThresholdDecryptionKey::new(&mut rng);
+        let tek = ThresholdEncryptionKey::new(&[d1.encryption_key(), d2.encryption_key()]);
+        let u = User::new(0, tek);
+
```

Comment:
maybe add a comment here about how we are NOT setting a match key. Took me a while to spot the difference with the test above.